### PR TITLE
[IMP] project: apply changes on Burndown chart and task kanban quickcreate

### PR DESCRIPTION
**Pull request by bastvdn**
Before this commit:

- The burndown chart displayed the stage in order of the data fetching which wasn't logical especial for a chart that supposed to represent the evolution of the tasks in the project
- When sorting all/my tasks by stage, It was possible to select any project in any quickcreate of the stages which didn't make much sense

After this commit:

The burndown chart legend is now ordered according to the stage sequence (previously was ordered randomly by comming data)
Modified the burndownChartModel, simply makes a RPC to get the stages and sequences then sort the legend elements (one by stage) with it
Display only the projects which uses the stage in the dropdown menu of the task kanban quickcreate (when grouping by stage)

When sorting by all/my tasks by stage, the quickcreate now only display the projects which contains the stage selected
done by adding a domain in the quickcreate form
(shoutout to LTU and AUON who actually found the fix)

Task-3067445

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    async fetchSequence() {
+        const activeId = this.model.metaData.context.active_id;
+        return await this.orm
+            .webSearchRead(
+                'project.task.type',
+                [['project_ids', 'in', activeId]],
+                ["name","sequence"]
+            ).then((data) => {return data.records})
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js:24]:
```suggestion
        const data = await this.orm.webSearchRead(
            "project.task.type",
            [["project_ids", "in", this.model.metaData.context.active_id]],
            ["name", "sequence"]
        );
        return data.records;
```

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                const { data } = chart;
+                let labels = data.datasets.map((dataset, index) => {
+                    let currentStage;
+                    let currentIndex;
+                    currentStage = this.stageSequence.find(item => item.name == dataset.label)
+                    if (currentStage == null){
+                        currentIndex = 1000
+                    } else {
+                        currentIndex = currentStage.sequence
+                    }
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js:46]:
```suggestion
                    const currentIndex = currentStage == null
                        ? 1000
                        : currentStage.sequence;
```
- also document the code when it is not obvious why
- 1000 is not the max sequence
- no matter what is solves, is it right to solve it by placing it last?

Comment (auon-odoo) [reply addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js:46]: currentSequence?

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                        lineDash: dataset.borderDash,
+                        lineDashOffset: dataset.borderDashOffset,
+                        lineJoin: dataset.borderJoinStyle,
+                        lineWidth: dataset.borderWidth,
+                        strokeStyle: dataset[referenceColor],
+                        pointStyle: dataset.pointStyle,
+                        datasetIndex: currentIndex,
+                    };
+                });
+                labels = labels.sort((a,b) => a.datasetIndex - b.datasetIndex)
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js:62]:
```suggestion
                labels = labels.sort((label) => label.datasetIndex);
```

File: addons/project/static/tests/burndown_chart_tests.js [hunk 4]
```diff
@@ -46,6 +46,15 @@ QUnit.module("Project", {}, () => {
...
                         ],
+                    },
+                    "project.task.type": {
+                        fields: {
+                            name: { string: "Name", type: "char" },
+                            sequence: { type: "integer"}
+
+                        },
+                        records: []
+
```
Comment (auon-odoo) [inline addons/project/static/tests/burndown_chart_tests.js:57]:
```suggestion
                    },
                    "project.task.type": {
                        fields: {
                            name: { string: "Name", type: "char" },
                            sequence: { type: "integer"}
                        },
                        records: []
```

Comment (auon-odoo) [reply addons/project/static/tests/burndown_chart_tests.js:57]: But sounds weird. Code what you undersand :laughing:

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js (hunk 3 above)
Comment (auon-odoo) [reply addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js:62]: or not?

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                [['project_ids', 'in', activeId]],
+                ["name","sequence"]
+            ).then((data) => {return data.records})
+    }
+
+    /**
+     * @returns {Object}
+     * @override
+    **/
+    getLegendOptions() {
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js:31]: weird things happen when clicking on a legend label.
it will be easier and cleaner to change the input rather hack the output

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Courage :)

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js (hunk 1 above)
Comment (bastvdn) [reply addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js]: Removed the renderer, added the sequence sorting in the burndown_chart_model

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js (hunk 2 above)
Comment (bastvdn) [reply addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js]: Removed the renderer, added the sequence sorting in the burndown_chart_model (no more currentIndex or currentSequence)

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js (hunk 3 above)
Comment (bastvdn) [reply addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js]: Removed the renderer, added the sequence sorting in the burndown_chart_model

File: addons/project/static/tests/burndown_chart_tests.js (hunk 4 above)
Comment (bastvdn) [reply addons/project/static/tests/burndown_chart_tests.js]: The 'serverData' object seems to become the backend data, when fetching data from the frontend in test, the data is fetched from this object.
So when i do the 
```
    async _fetchStageSeq(searchParams) {
        const data = await this.orm.webSearchRead(
            "project.task.type",
            [["project_ids", "in", searchParams.context.active_id]],
            ["name", "sequence"]
        );
```
The serverData needs to have some 'project.task.type' key.
The problem is that i would like it to be the burndown_chart.stage_id but i didn't find how to do this, i need help on this one ^^'
When fetching the points in the _graph_model.js_
```
async _loadDataPoints(metaData) {
...
this.orm
  .webReadGroup(
      resModel,
      domain.arrayRepr,
      measures,
      groupBy.map((gb) => gb.spec),
      {
          lazy: false, // what is this thing???
          context: { fill_temporal: true, ...this.searchParams.context },
      }
  )
...
}
```

the 'resModel' is changed from 'project.task.burndown.chart.report' ( in production ) to 'burndown_chart' ( in testing ). 
This is done somewhere in the 
```
await makeView(makeViewParams);
```
in _burndown_chart_tests.js_ but i couldn't find how, tell me if you have some insights about this

File: addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js (hunk 5 above)
Comment (bastvdn) [reply addons/project/static/src/views/burndown_chart/burndown_chart_renderer.js]: Removed the renderer, added the sequence sorting in the burndown_chart_model

**Comment by bastvdn** [comment]
Hey @auon-odoo, thanks a lot for the review, i fixed the problems you found but i still need some help with the tests.
Sorry i took that much time to fix my PR, tell me if anything is wrong in my correction :)

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js [hunk 6]
```diff
@@ -1,14 +1,147 @@
...
+                    domains: new Array(dataLength).fill([]),
+                    label: datasetLabel,
+                    originIndex: originIndex,
+                };
+            }
+            datasetsTmp[datasetLabel].data[labelIndex] = value;
+            datasetsTmp[datasetLabel].domains[labelIndex] = domain;
+            datasetsTmp[datasetLabel].trueLabels[labelIndex] = trueLabel;
+            datasetsTmp[datasetLabel].orderIndex = this.stageSeq[datasetLabel] || 0;
+        }
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_model.js:112]: We should isolate this part of the code in a method. This way, we'ill just add `orderIndex` to its result.

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js [hunk 7]
```diff
@@ -1,14 +1,147 @@
...
+        return { datasets, labels };
+    }
+
     /**
      * @protected
      * @override
      */
     async _loadDataPoints(metaData) {
-        metaData.measures.__count.string = this.env._t('# of Tasks');
+        metaData.measures.__count.string = this.env._t("# of Tasks");
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_model.js:144]: Unnecessary change :(

File: addons/project/models/project.py [hunk 8]
```diff
@@ -749,6 +749,12 @@ def _mail_get_message_subtypes(self):
     def action_project_task_burndown_chart_report(self):
         action = self.env['ir.actions.act_window']._for_xml_id('project.action_project_task_burndown_chart_report')
         action['display_name'] = _("%(name)s's Burndown Chart", name=self.name)
+        context = action['context'].replace('active_id', str(self.id))
+        context = ast.literal_eval(context)
+        context.update({
+            'stage_seq': dict([(stage.name, stage.sequence) for stage in self.type_ids])
```
Comment (auon-odoo) [inline addons/project/models/project.py:755]:
```suggestion
            'stage_seq': {stage.name: stage.sequence for stage in self.type_ids}
```
Isn't that the same?

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thank you! :) Let's clear that out a bit

File: addons/web/static/src/views/graph/graph_model.js [hunk 9]
```diff
@@ -243,6 +243,7 @@ export class GraphModel extends Model {
                     domains: new Array(dataLength).fill([]),
                     label: datasetLabel,
                     originIndex: originIndex,
+                    identifier: identifier,
```
Comment (Polymorphe57) [inline addons/web/static/src/views/graph/graph_model.js:246]: `identifier,`

Comment (Polymorphe57) [reply addons/web/static/src/views/graph/graph_model.js:246]: don't think it is good, two data points with different identifiers can have same datasetLabel

Comment (bastvdn) [reply addons/web/static/src/views/graph/graph_model.js]: @Polymorphe57 Hello again, i wasn't able to finish this PR in the last few months sorry :)
Could you tell me if my new fix is accurate?
Thanks for the review

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js [hunk 10]
```diff
@@ -1,14 +1,78 @@
...
+            "project.task.type",
+            [["project_ids", "in", searchParams.context.active_id]],
+            ["name", "sequence", "id"]
+        );
+
+        const dataClean = {};
+        data.records.forEach((obj) => {
+            dataClean[obj.id] = { sequence: obj.sequence, name: obj.name };
+        });
+        return dataClean;
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_model.js:33]: Possible to map here :slightly_smiling_face:

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js [hunk 11]
```diff
@@ -1,14 +1,78 @@
...
+            (ds) => {
+                // In case there was no identifiers in the datasets 
+                if (!ds.identifiers || ds.identifiers.size === 0) {
+                    ds.orderIndex = -1;
+                }
+                else {
+                    // In case the identifier doesn't correspond to any stage.id
+                    ds.orderIndex = this.stageSeq[Math.min(...ds.identifiers)] !== undefined ? this.stageSeq[Math.min(...ds.identifiers)].sequence : -1;
+                }
+            });
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_model.js:66]:
```suggestion
        for (const dataset in datasets) {
            if (!dataset.identifiers || !dataset.identifiers.size) {
                // In case there was no identifiers in the datasets 
                dataset.orderIndex = -1;
            } else {
                // In case the identifier doesn't correspond to any stage.id
                const stage = this.stageSeq[Math.min(...dataset.identifiers)];
                dataset.orderIndex = stage !== undefined ? stage.sequence : -1;
            }
        });
```
- This way to iterate is preferred. (Or map if possible ?)
- `blabla.size === 0` is the same as `!blabla.size`.
- `} else {`
- `this.stageSeq[Math.min(...dataset.identifiers)]` is :vomiting_face:, let's write it only once.
- Comments on the same level.

File: addons/project/models/project.py [hunk 12]
```diff
@@ -760,6 +760,12 @@ def _mail_get_message_subtypes(self):
     def action_project_task_burndown_chart_report(self):
         action = self.env['ir.actions.act_window']._for_xml_id('project.action_project_task_burndown_chart_report')
         action['display_name'] = _("%(name)s's Burndown Chart", name=self.name)
+        context = action['context'].replace('active_id', str(self.id))
+        context = ast.literal_eval(context)
+        context.update({
+            'stage_seq': {stage.id: {'sequence': stage.sequence, 'name': stage.name} for stage in self.type_ids}
+            })
```
Comment (auon-odoo) [inline addons/project/models/project.py:767]:
```suggestion
        context.update({
            'stage_name_and_sequence_per_id': {
                stage.id: {
                    'sequence': stage.sequence,
                    'name': stage.name
                } for stage in self.type_ids
            }
        })
```
- Close the brackets you opened at same indent.
- Name the dict more explicitly.
- Unwrap it totally.

File: addons/project/models/project.py [hunk 13]
```diff
@@ -966,6 +972,7 @@ def _get_stat_buttons(self):
                 'action': 'project.action_project_task_burndown_chart_report',
                 'additional_context': json.dumps({
                     'active_id': self.id,
+                    'stage_seq': {stage.id: {'sequence': stage.sequence, 'name': stage.name} for stage in self.type_ids}
```
Comment (auon-odoo) [inline addons/project/models/project.py:975]: Same

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js (hunk 10 above)
Comment (auon-odoo) [reply addons/project/static/src/views/burndown_chart/burndown_chart_model.js:33]: Call your variable `res` or `stageSequenceAndNamePerId`

Comment (auon-odoo) [reply addons/project/static/src/views/burndown_chart/burndown_chart_model.js:33]: And you method name `_fetchStageInfo` or `fetchStageSequenceAndNamePerId`.

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js [hunk 14]
```diff
@@ -1,14 +1,78 @@
...
+    }
+
+    /**
+     * @param {SearchParams} searchParams
+     */
+    async load(searchParams) {
+        this.searchParams = searchParams;
+        // if the stage_seq wasn't given by the action (for exemple if the page is simply reloaded)
+        if (!this.searchParams.context.stage_seq) {
+            this.stageSeq = await this._fetchStageSeq(searchParams);
```
Comment (auon-odoo) [inline addons/project/static/src/views/burndown_chart/burndown_chart_model.js:43]: `this.stageInfo`, `this.stageData`, or even better `stageSequenceAndNamePerId`.
Point being that, when I read your code, I immediately understand what your doing when you write `this.stageSeq[blabla]` :zap:

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thank you!! Some comments left...

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js (hunk 11 above)
Comment (auon-odoo) [reply addons/project/static/src/views/burndown_chart/burndown_chart_model.js:66]: Also, min is pretty arbitrary... But I guess, it has to be since we have a Set.
@Polymorphe57 Thank you for your feedback! I'm not sure I understand your point. Could explain a little further (maybe wwith a usecase) please? :pray:

**Review by Polymorphe57 — COMMENTED** [review]
I have left some comments and push a commit with some changes. You can keep it or not. It's up to you

**Comment by Polymorphe57** [comment]
just one more question. Is there no way to ask to read_group to sort correctly the groups directly via orderby?

File: addons/project/models/project.py (hunk 13 above)
Comment (bastvdn) [reply addons/project/models/project.py:975]: DOne

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js (hunk 14 above)
Comment (bastvdn) [reply addons/project/static/src/views/burndown_chart/burndown_chart_model.js]: Done

File: addons/web/static/src/views/graph/graph_model.js [hunk 15]
```diff
@@ -409,6 +411,7 @@ export class GraphModel extends Model {
                         for (const group of data.groups) {
                             const { __domain, __count } = group;
                             const labels = [];
+                            let identifier;
```
[the remark this answers is not in the document]
Comment (bastvdn) [reply addons/web/static/src/views/graph/graph_model.js]: Ok The way you're fetching the whole groupby field_name and spec is more thorough, i'll use this. Thanks

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js [hunk 16]
```diff
@@ -1,14 +1,78 @@
...
+
+    /**
+     * Fetch the sequence of each stage in the project. This function alters this.stageSeq
+     * @protected
+     * @param {Object} searchParams
+     */
+    async _fetchStageSeq(searchParams) {
+        const data = await this.orm.webSearchRead(
+            "project.task.type",
+            [["project_ids", "in", searchParams.context.active_id]],
```
[the remark this answers is not in the document]
Comment (bastvdn) [reply addons/project/static/src/views/burndown_chart/burndown_chart_model.js]: We always access the burndown chart through a project, so the active_id should always be defined. Tried to remove it by force through the URL and the burndown chart shouldn't work in this case.

As for the filter, yes when you remove it it'll fetch every stage along the database to create the burndown chart, i'll add this possibility to my corner case, if the context is missing the *default_project_id* i'll fetch the sequence of every stage, so the burndown chart will still be sorted correctly according to the sequence of every stage

**Review by auon-odoo — APPROVED** [review]
Lgtm :fire:

File: addons/project/models/project_task.py [hunk 17]
```diff
@@ -931,6 +931,9 @@ def create(self, vals_list):
                         default_project_id=project_id
                     ).default_get(['stage_id']).get('stage_id')
                 vals["stage_id"] = default_stage[project_id]
+                new_stage = self.env['project.task.type'].browse(vals['stage_id'])
+                if project_id not in new_stage.project_ids.ids:
+                    new_stage.write({'project_ids': [Command.link(project_id)]})
```
Comment (xavierbol) [inline addons/project/models/project_task.py:936]: Alternatively, I think we could do something like this when we create a project in project_id many2one field in quick create of the kanban. 🤔 

```diff
diff --git a/addons/project/models/project_task.py b/addons/project/models/project_task.py
index 7e2bbd67ed75..b0e547fdd937 100644
--- a/addons/project/models/project_task.py
+++ b/addons/project/models/project_task.py
@@ -905,9 +905,6 @@ class Task(models.Model):
                         default_project_id=project_id
                     ).default_get(['stage_id']).get('stage_id')
                 vals["stage_id"] = default_stage[project_id]
-                new_stage = self.env['project.task.type'].browse(vals['stage_id'])
-                if project_id not in new_stage.project_ids.ids:
-                    new_stage.write({'project_ids': [Command.link(project_id)]})
             # user_ids change: update date_assign
             if vals.get('user_ids'):
                 vals['date_assign'] = fields.Datetime.now()
diff --git a/addons/project/views/project_task_views.xml b/addons/project/views/project_task_views.xml
index 8d1348995848..5e46db7671fe 100644
--- a/addons/project/views/project_task_views.xml
+++ b/addons/project/views/project_task_views.xml
@@ -495,7 +495,7 @@
                     <group>
                         <field name="display_name" string= "Task Title" placeholder="e.g. Send Invitations"/>
                         <field name="project_id" widget="project_private_task" invisible="context.get('default_project_id', False)" placeholder="Private" class="o_project_task_project_field"
-                        domain="[('type_ids', 'in', context['default_stage_id'])] if context.get('default_stage_id') else []"/>
+                        domain="[('type_ids', 'in', context['default_stage_id'])] if context.get('default_stage_id') else []" context="{'default_type_ids': [(4, context.get('default_stage_id', False))]}"/>
                         <field name="user_ids" options="{'no_open': True, 'no_quick_create': True}" domain="[('share', '=', False), ('active', '=', True)]"
                             widget="many2many_avatar_user"/>
                         <field name="company_id" invisible="1"/>
```

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂 

You can squash your commits into one. 🙂

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js [hunk 18]
```diff
@@ -1,14 +1,84 @@
...
+     */
+    async load(searchParams) {
+        const { context, groupBy } = searchParams;
+
+        if (groupBy.includes("stage_id")) {
+            if (context.stage_name_and_sequence_per_id && context.default_project_id) {
+                this.stageSeqAndNamePerId = context.stage_name_and_sequence_per_id;
+            } else {
+                // if the stage_name_and_sequence_per_id wasn't given by the action (for example if the page is simply reloaded)
+                this.stageSeqAndNamePerId = await this._fetchStageInfo(context);
```
[the remark this answers is not in the document]
Comment (bastvdn) [reply addons/project/static/src/views/burndown_chart/burndown_chart_model.js:48]: I prefer not to do an additional RPC if the infos about the stage sequence can be loaded in the action context :shrug: Do you think it's better to be consistent and always do the RPC?

File: addons/project/models/project_task.py (hunk 17 above)
Comment (bastvdn) [reply addons/project/models/project_task.py]: it works :-)

File: addons/project/static/src/views/burndown_chart/burndown_chart_model.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/project/static/src/views/burndown_chart/burndown_chart_model.js:48]: In general we avoid filling the context because if the action was defined in xml only, we cannot really do that and so we will have some inconsistency.

But here, it could be acceptable since the report is costly. 🙂

**Review by xavierbol — APPROVED** [review]
LGTM, thanks again for your work. 🙂 

robodoo r+

Thanks @Polymorphe57 for your help and your review 🙂 

Have a nice day! 😁
